### PR TITLE
[WIP] Basic Qt GUI

### PR DESCRIPTION
**Pull request by spxtr**
I'm attempting to lay the foundation for a new DolphinQt GUI using idiomatic Qt constructs. Here's where I'm at so far:

* GameTracker watches directories for new/edited files matching the usual filters. If a game file shows up or is modified it loads it in a worker thread (so that the GUI doesn't hang when you add a directory with lots of games in it), then passes it up to GameListModel. I haven't implemented loading GameFiles from cache yet, but that's not too tough.
* GameListModel keeps a list of GameFiles and exposes them as a Qt Model. This is actually a super-general, ugly interface, but it lets the Views happen automatically, which is nice.
* GameList is a stack of two Views, a TableView like in DolphinWX and a ListView that shows big icons. I haven't done a TreeView, but it's not too tough either, just need to make a ProxyModel.
* MainWindow ties the whole room together.
* I'm not sure if Host is done right, but I think it's the right idea. It's a singleton that talks to MainWindow exclusively through signals and slots so that there can't be any race condition.
* I did Resources in just about the same was as the current DolphinQt does it, architecturally speaking. Not sure if that's optimal.

Style note because I've seen Dolphin people make this mistake: When you make QObject-derived-things, if you pass a parent to the constructor then they'll automatically be deleted when the parent is deleted. It's an error to delete a QObject-derived-thing that might have events pending, and it's  and error to use a smart pointer for parented objects.

Anyway, I'm very interested in style comments and architecture comments. I'm well aware that various functions are unimplemented or half-implemented. The point is to use this as a platform to make design decisions and prevent the rise of DolphinWX2.

screenshot
screenshot2
screenshot3

### CODE REVIEW

**Comment by shuffle2** [comment]
please fix the linter errors, not sure what's up with the windows bot...

**Comment by spxtr** [comment]
The linter isn't going to be happy because it's normal to separate Qt includes from standard C++ includes. Here I include <map>, <QStuff>, and <string>. The linter wants me to do it in that order, where normally I would do <map>, <string>, and then all the <QStuff>.

Except one of the other includes is out of order, I'll fix that for sure.

It won't compile on windows, that's expected. I'll get that working soon.

**Comment by shuffle2** [comment]
sure but the buildbot logs show successful build returning 1 (error).

**Comment by shuffle2** [comment]
also, Just Make The Linter Happy.

**Comment by spxtr** [comment]
Ok

File: Source/Core/DolphinQt/GameFile.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+#include <string>
+
+#include "GameFile.h"
+#include "Common/FileUtil.h"
+#include "Core/ConfigManager.h"
+#include "DiscIO/Volume.h"
+#include "DiscIO/VolumeCreator.h"
+
+
+GameFile::GameFile(const QString& path)
```
Comment (lioncash) [inline Source/Core/DolphinQt/GameFile.cpp]: This constructor should be broken out into initialization functions.

File: Source/Core/DolphinQt/GameFile.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	: m_path(path)
+{
+	// FIXME Do this whole thing better.
+	DiscIO::IVolume* v = DiscIO::CreateVolumeFromFilename(m_path.toStdString());
```
Comment (lioncash) [inline Source/Core/DolphinQt/GameFile.cpp]: this can be a unique_ptr.

File: Source/Core/DolphinQt/GameFile.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#include <map>
+#include <QDebug>
+#include <QImage>
+#include <QObject>
+#include <QString>
+#include <QStringList>
+#include <QStringListIterator>
+#include <string>
+
+#include "GameFile.h"
```
Comment (lioncash) [inline Source/Core/DolphinQt/GameFile.cpp]: This should be a relative path like the others.

File: Source/Core/DolphinQt/MainWindow.cpp [hunk 4]
```diff
@@ -1,362 +1,302 @@
...
 }
 
-void DMainWindow::DoStartPause()
+void MainWindow::PopulateGameTable()
 {
-	if (Core::GetState() == Core::CORE_RUN)
+	m_game_table->setEnabled(false);
+	while (m_game_table->rowCount() > 0)
+		m_game_table->removeRow(0);
+	m_games.clear();
```
Comment (lioncash) [inline Source/Core/DolphinQt/MainWindow.cpp]: This bit of code (from the beginning of the function to here) should be extracted into its own function (ClearGameTable or some better name)

File: Source/Core/DolphinQt/MainWindow.cpp [hunk 5]
```diff
@@ -1,362 +1,302 @@
...
-	return true;
-}
-
-void DMainWindow::OnReset()
-{
-	// TODO: Movie needs to be reset here
-	ProcessorInterface::ResetButton_Tap();
+	{
+		QModelIndexList indices = m_game_table->selectionModel()->selectedRows();
+		if (indices.count() > 0) {
```
Comment (lioncash) [inline Source/Core/DolphinQt/MainWindow.cpp]: Brace on new line

**Comment by JosJuice** [comment]
Wait, what? Is the current Qt GUI going to be replaced? (cc @waddlesplash)

**Comment by waddlesplash** [comment]
Um, yeah, what? You got rid of a lot of features for no reason, renamed classes that had been already agreed upon, readded stuff we'd already agreed to remove (e.g. refresh button) and switched from .ui to manually initialized GUIs.

I'm sorry, I don't see how this is anything but 10 steps back for DolphinQt.

File: Source/Core/DolphinQt/Host.cpp [hunk 6]
```diff
@@ -1,30 +1,19 @@
...
-void Host_UpdateMainFrame()
-{
-	// TODO
-}
+void Host_UpdateMainFrame() {}
 
 void Host_UpdateTitle(const std::string& title)
 {
-	// TODO
+	g_main_window->setWindowTitle(QString::fromStdString(title));
```
Comment (waddlesplash) [inline Source/Core/DolphinQt/Host.cpp]: This is illegal, because `g_main_window` is on a different thread than the thread that calls the `Host_*` functions, so without locking this is illegal. And because Qt does all sorts of stuff internally that are unlockable, the only way to do this is with events. Which I already did in #3032.

File: Source/Core/DolphinQt/Host.cpp [hunk 7]
```diff
@@ -32,15 +21,11 @@ void* Host_GetRenderHandle()
...
-{
-	DRenderWidget* render_widget = g_main_window->GetRenderWidget();
-	qApp->postEvent(render_widget, new QResizeEvent(QSize(w, h), render_widget->size()));
-}
+void Host_RequestRenderWindowSize(int w, int h) {}
 
 bool Host_RendererHasFocus()
 {
-	return g_main_window->RenderWidgetHasFocus();
+	return g_main_window->GetRenderWidget()->isActiveWindow();
```
Comment (waddlesplash) [inline Source/Core/DolphinQt/Host.cpp]: This doesn't always return true if the RenderWidget is a child of the main window instead of its own window, IME. I could be misremembering that, though.

File: Source/Core/DolphinQt/Main.cpp [hunk 8]
```diff
@@ -1,65 +1,18 @@
...
-{
-#ifdef Q_OS_OSX
-	return SL("Mac OS X 10.9");
-#elif defined(Q_OS_WIN)
-	return SL("Windows Vista SP2");
-#else
-	return SL("Unknown");
-#endif
-}
-
```
Comment (waddlesplash) [inline Source/Core/DolphinQt/Main.cpp]: Why did you delete this code? It's still needed.

**Comment by waddlesplash** [comment]
Okay, full critique. Here goes: 

> 11:18 PM <spxtr> Well whoever made those icons in Resources, shame on you. That's a silly thing to do.

If you had actually looked at the Resources class before deleting it, you would've noticed it's there for 2 reasons:
1. Properly loading HiDPI (`@2x`) vs. non-HiDPI resources on appropriate systems
2. Reloading all theme-based icons/pixmaps when a theme changes, and signaling everything who uses them.

>1:23 AM <spxtr> For the most part, at least.
>1:24 AM <endrift> any particular reason for that?
>1:25 AM <spxtr> The WX code is pretty horrific. I don't want to fall into the trap of adding bits and pieces until it turns out like that.

The current DolphinQt reuses almost no code from WX, save in GameFile where WX is mostly sane so it wasn't an issue. And you'll be adding bits and pieces here anyway. Creating something incrementally is not a reason it'll turn out like WX.

> 1:39 AM <spxtr> Also I'm not a big fan of using designer files for something this simple.
> 1:40 AM <spxtr> I find that, for instance, MainWindow.ui in the current Qt version is very difficult to understand at a glance.
> 1:40 AM <spxtr> And I don't want to break out QtCreator.

So use Qt Designer, it comes with libqt4-dev on Debian systems (and similarly everywhere else). UI files provide:
 * Stuffing long initialization logic somewhere where you don't have to look at it and let it clog up your main source files
 * Creating the nice `retranslateUi` function, which resets all strings on the current UI, so we can switch languages at runtime without a restart
 * Handling connection of simple signals/slots

You can put all this in a `.h` file as well, sure. But why not let `uic` do all the heavy lifting for you in that case?

You also deleted a lot of platform-specific stuff (e.g. .icns, Mac bundling stuff, etc.) for no apparent reason, as well as the About and SystemInfo dialogs. Why? What was wrong with these that made them too "WX-like"?

Your new GameFile class also is missing a lot of features (Homebrew XML loading, cache support, proper decoding on a lot of parameters) and looks like another complete rewrite. Why?

You also got rid of *all* the GameList code, with only a comment in IRC about how it "was nice". I realize this code doesn't exactly follow Qt paradigms (no use of models), that wasn't exactly possible because of the GameTree feature (which multiple people, including @JMC47, really liked). The GameList also watched the filesystem for changes, and doing incremental updates from that, instead of requiring people to manually hit a "refresh" button which re-scans the whole directory tree (and can take a long time). Not to mention the loss of the other GameList styles.

**Comment by waddlesplash** [comment]
**TL;DR**: This code has a vastly different architecture, gets rid of most of the nice internal cleanups we were able to do from WX (filesystem watching, UIs), gets rid of most of the features DQt had, and has illegal API calls without proper locking (which is impossible). I don't get it.

**Comment by neobrain** [comment]
Mind providing an actual summary of what this PR adds/changes?

**Comment by neobrain** [comment]
@waddlesplash To be fair, .ui file are terrible for even the most trivial GUIs as soon as you try to do anything remotely dynamic. There is no way to get around coding large parts of the GUI without .ui files at all, and then you might as well write all of it manually instead of fragmenting the GUI code in two different designing methods.

Btw, I think you might not be the best person to complain about vast architecture changes and features removals from DQt.

**Comment by waddlesplash** [comment]
> There is no way to get around coding large parts of the GUI without .ui files at all, and then you might as well write all of it manually instead of fragmenting the GUI code in two different designing methods.

What dynamic UI is there in Dolphin? Config panels are all that I'm aware of, and that's nothing that a QStackedWidget + 10 manually written LoC won't take care of.

**Comment by spxtr** [comment]
Thanks for the comments, @waddlesplash. You're right that this is a regression in several ways. I wan't clear enough in my intentions here: it's a rewrite from scratch, not just a refactor or something like that. I should likely have put it in its own separate folder. I'll make a more complete response to your comments tonight.

In the meantime, does anyone know what changes I need to make to get buildbot to upload the DolphinQt executable? I see that it compiles it but then doesn't include it in the upload.

**Comment by waddlesplash** [comment]
> it's a rewrite from scratch, not just a refactor or something like that

Yeah, why?

**Comment by shuffle2** [comment]
@spxtr https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/DolphinQt/DolphinQt.vcxproj#L179
remove `Condition="'$(I_AM_BUILDACUS)'==''"` ;)
Note: Make sure you put it back before merging PR to master.

**Comment by spxtr** [comment]
@waddlesplash et al:

Switching to Qt will allow for some major improvements to Dolphin that have nothing to do with the front-end. The majority of Common can be dropped in favor of Qt counterparts that are simply better. Improvements include:

* Drop the refresh button in favor of a QFileSystemWatcher.
* Drop Ini files and the SConfig monstrosity in favor of QSettings.
* Drop MathUtil for various Qt classes and functions.
* Drop FileUtil for a combination of QFile and QDir.
* Drop FileSearch in favor of QDirIterator.
* Drop Threads and Timers in favor of QThreads and QTimers.

and so on. I'm not sure whether or not we should maintain a framework-independent interface or just drop Qt stuff into the Core, but I'm leaning towards the latter. For just about all of these the Qt version will be more portable, faster, and native than Dolphin's. Unfortunately, these reach out of the DolphinQt directory, and not all of the extant solutions are well-defined interfaces that we can just drop Qt into. In particular, SConfig is a mess and will be a pain to fix.

In this WIP branch I want to start implementing these sorts of changes. Starting from scratch allows us to keep the two worlds separate from the get go. If we're going to replace SConfig with QSettings, for instance, we'd like to avoid building a complete GUI using SConfig, then having to switch it over.

I hope I communicated that better this time. The goal is to have a simple GUI that I can use as a test platform for architectural changes that will be coming soon. I really should've put it in a separate folder, but I wanted to hijack some of the existing infrastructure.

**Comment by delroth** [comment]
> The majority of Common can be dropped in favor of Qt counterparts that are simply better. 

No. Nothing outside of the DolphinQt directory will depend on Qt, and we will keep the nogui build without any dependency on Qt. Just like we do nowadays with WX.

Whatever you're dreaming about won't happen.

**Comment by delroth** [comment]
And this is not a negotiable point, btw. No dependency on huge libraries like Qt in Common or Core has been a rule enforced since at least 2008, and it has brought us a lot of good.

**Comment by spxtr** [comment]
It's not too difficult to use a framework-independent interface if there's a policy of keeping Qt out of the core.

**Comment by waddlesplash** [comment]
> And this is not a negotiable point, btw. No dependency on huge libraries like Qt in Common or Core has been a rule enforced since at least 2008, and it has brought us a lot of good.

+1 from me. DolphinQt already does a lot of the other things @spxtr mentioned, like:

> Drop the refresh button in favor of a QFileSystemWatcher.

[It's already been doing this for a while](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/DolphinQt/GameList/GameTracker.cpp#L28).

> For just about all of these the Qt version will be more portable, faster, and native than Dolphin's. 

LOLNO. Qt tries to interact with native too much. It's really good, but it's also pretty slow.

>SConfig is a mess and will be a pain to fix.

Agreed but there are already docs for switching to an onion-type config system, nobody just implemented it yet. This has nothing to do with Qt; QSettings deserves to die in a fire already IMO since it's impossible to hand-edit due to nobody knowing where/how it stores things and having a weird format for doing so.

Also the rest of those complains have 0 to do with DolphinQt's internal architecture and don't explain why you started from scratch.

**Comment by waddlesplash** [comment]
> It's not too difficult to use a framework-independent interface if there's a policy of keeping Qt out of the core.

Uh, no? There's a lot of logic you'd have to duplicate for no gain and a lot of effort.

**Comment by JosJuice** [comment]
> In this WIP branch I want to start implementing these sorts of changes. Starting from scratch allows us to keep the two worlds separate from the get go. If we're going to replace SConfig with QSettings, for instance, we'd like to avoid building a complete GUI using SConfig, then having to switch it over.

The current DolphinQt is not using SConfig much. It does use it for getting emulation ratings and game list settings, but yours does that too. I don't see what the point is in rewriting DolphinQt, even if we are to ignore delroth's very good point.

**Comment by spxtr** [comment]
Quite a few changes, I'll update the description once I fix the windows compilation issues.

File: Source/Core/DolphinQt/GameList/GameFile.h [hunk 9]
```diff
@@ -1,95 +1,93 @@
...
-	// 0 is the first disc, 1 is the second disc
-	u8 GetDiscNumber() const { return m_disc_number; }
-	const QPixmap GetBitmap() const
+
+	QString GetLongName() const { return GetLongName(GetLanguage()); }
+	QString GetLongName(DiscIO::IVolume::ELanguage lang) const
 	{
-		if (m_banner.isNull())
-			return Resources::GetPixmap(Resources::BANNER_MISSING);
+		return m_long_names[lang];
```
Comment (JosJuice) [inline Source/Core/DolphinQt/GameList/GameFile.h]: `m_long_names` doesn't always contain the language you want. This won't return a name if e.g. the system language is set to English and the volume is a Japanese GameCube disc.

Comment (JosJuice) [inline Source/Core/DolphinQt/GameList/GameFile.h]: GetLongName(ELanguage) returning an empty string if the language is missing might be fine, but it's not fine for GetLongName() since that's what the game list uses to get a name to display.

Comment (spxtr) [inline Source/Core/DolphinQt/GameList/GameFile.h]: Ah, I see that's handled in DolphinWX/ISOFile.cpp. Not a hard fix. Thanks :)

Comment (waddlesplash) [inline Source/Core/DolphinQt/GameList/GameFile.h]: Uh, ISOFile is kinda messy. Why didn't you just stick with the original GameFile?

**Comment by waddlesplash** [comment]
You still have a complete rearchitecture of DolphinQt for no reason. There were a lot of advances you just threw away, features you threw away (GameList cache, treed gamelist, etc.), and other general changes I don't agree with. I still don't understand what the point is.

**Comment by waddlesplash** [comment]
> The point is to use this as a platform to make design decisions and prevent the rise of DolphinWX2.

I've never heard anyone suggest DolphinWX2. The systematic cleanups of DolphinWX (mostly by @lioncash) have slowed down; nobody is planning a complete DolphinWX rewrite. DolphinQt has already been a platform to make design decisions.

> When you make QObject-derived-things, if you pass a parent to the constructor then they'll automatically be deleted when the parent is deleted. It's an error to delete a QObject-derived-thing that might have events pending, and it's and error to use a smart pointer for parented objects.

Uh, what?! From the QObject docs:

> You can also delete child objects yourself, and they will remove themselves from their parents

Do you even know what you're talking about?!?

File: Source/Core/DolphinQt/GameList/GameFile.h (hunk 9 above)
Comment (JosJuice) [inline Source/Core/DolphinQt/GameList/GameFile.h]: If you want to rewrite it, ISOFile is about as good of a reference as the original GameFile, except for some notably wx-specific parts like caching and image loading. The question is why it needs to rewritten, though...

**Comment by JosJuice** [comment]
If being "a platform to make design decisions" is the point of this PR, could you tell us about those design decisions? The PR description mentions what you've done, but not why you're changing things, what you're improving, and most importantly, why a complete rewrite is needed.

**Comment by spxtr** [comment]
In a vacuum, sure, you can delete a parented object, but I said it's an error to delete an object that might have pending events.

> Warning: Deleting a QObject while pending events are waiting to be delivered can cause a crash. You must not delete the QObject directly if it exists in a different thread than the one currently executing. Use deleteLater() instead, which will cause the event loop to delete the object after all pending events have been delivered to it.

Using a smart pointer for parented items is a double free when the parent is destroyed before the  smart pointer does its thing.

> Do you even know what you're talking about?!?

That's not a nice thing to say.

**Comment by endrift** [comment]
> it's and error to use a smart pointer for parented objects.

I'm under the impression this is actually true, in case the parent deletes it out from under the smart pointer.

**Comment by waddlesplash** [comment]
> In a vacuum, sure, you can delete a parented object, but I said it's an error to delete an object that might have pending events.

OK, so I need to rework the RenderWidget memory management. ~10 LoC to change.

> Using a smart pointer for parented items is a double free when the parent is destroyed before the smart pointer does its thing.

Except the only use of smart pointers is inside the class which owns the thing the smart pointer points to. And aside from RenderWidget, IIRC the only other things are UIC classes, which don't destroy their children on delete anyway.

None of this is anywhere near a reason for a rewrite. Even getting rid of 100% of the smart pointers would take me only 30m or so.

**Comment by spxtr** [comment]
I'll rebase later tonight. I made the long name language fix, made it open in a new window or fullscreen if needed, and changed how it talks to Host a bit.

Anyone have experience with Qt's UI testing facilities? I'll try to figure them out later, as I'd like to get this tested well before doing more. Any help appreciated.
